timemachine/md/hrex.py

```diff
@@ -0,0 +1,247 @@
+    get_log_q_fn: Callable[[List[_Replica]], Callable[[ReplicaIdx, StateIdx], float]],
+    n_samples: int,
+    n_samples_per_iter: int,
+    n_swap_attempts_per_iter: Optional[int] = None,
+) -> Tuple[List[List[_Samples]], HrexDiagnostics]:
+    """Sample from a sequence of states using Hamiltonian Replica EXchange (HREX).
+
+    This implementation uses a method described in [1] (in section III.B.2) to generate effectively uncorrelated
+    permutations by attempting many consecutive nearest-neighbor swap moves. By default, the number of swap moves is
+    determined as a function of the number of states (:math:`K`) as :math`N_{\text{swaps}} = K^4`, a heuristic also
+    described in [1].
+
+    References
+    ----------
+    [1]: http://dx.doi.org/10.1063/1.3660669
+
+    Parameters
+    ----------
+    replicas: sequence of _Replica
+        Sequence of initial states of each replica
+
+    sample_replica: (_Replica, StateIdx, n_samples: int) -> _Samples
+        Local sampling function. Should return n_samples samples from the given replica and state
+
+    replica_from_samples: _Samples -> _Replica
+        Function that returns a replica state given a sequence of local samples. This is used to update the state of
+        individual replicas following local sampling.
+
+    neighbor_pairs: sequence of (StateIdx, StateIdx)
+        Pairs of states for which to attempt swap moves
+
+    get_log_q_fn: sequence of _Replica -> ((ReplicaIdx, StateIdx) -> float)
+        Function that returns a function from replica-state pairs to log unnormalized probability. Note that this is
+        equivalent to the simpler signature (_Replica, StateIdx) -> float; the "curried" form here is to allow for the
+        implementation to compute the full matrix as a batch operation when this is more efficient.
+
+    n_samples: int
+        Total number of local samples (e.g. MD frames)
+
+    n_samples_per_iter: int
+        Number of local samples (e.g. MD frames) per HREX iteration
+
+    n_swap_attempts_per_iter: int or None, optional
+        Number of neighbor swaps to attempt per iteration. Default is given by :py:func:`get_swap_attempts_per_iter_heuristic`.
+
+    Returns
+    -------
+    List[List[_Samples]]
+        samples grouped by state and iteration
+
+    HrexDiagnostics
+        HREX statistics (e.g. swap rates, replica-state distribution)
+    """
+
+    n_replicas = len(replicas)
+
+    if n_swap_attempts_per_iter is None:
+        n_swap_attempts_per_iter = get_swap_attempts_per_iter_heuristic(n_replicas)
+
+    hrex = Hrex.from_replicas(replicas)
+
+    samples_by_state_by_iter: List[List[_Samples]] = []
+    replica_idx_by_state_by_iter: List[List[ReplicaIdx]] = []
+    fraction_accepted_by_pair_by_iter: List[List[Tuple[int, int]]] = []
+
+    for n_samples_batch in batches(n_samples, n_samples_per_iter):
+        sample_replica_ = lambda replica, state_idx: sample_replica(replica, state_idx, n_samples_batch)
```

Comment:
Is there somewhere that asserts that the number of samples returned by `sample_replica` equals `n_samples_batch`?

Reply:
Not currently. I'm not sure that makes sense to add here, since this function is generic in `_Samples` (e.g. it could be a list or a tuple of lists, as is the case in RBFE simulations), so it doesn't know how to query the length of the returned samples.

Reply:
The `n_samples_batch` argument is currently used to handle cases where the total number of samples requested is not a multiple of `n_samples_per_iter`. In this case, the last batch will have fewer samples. We could consider just "rounding up" `n_samples` to the nearest multiple of `n_samples_per_iter` (which would allow removing this argument), but I wasn't immediately sure of the best approach here.

Reply:
Right, missed that. Thanks for the clarification